pkg/controller/shoot/shoot_quota_control.go

```diff
@@ -0,0 +1,195 @@
+	return &defaultQuotaControl{k8sGardenClient, k8sGardenInformers}
+}
+
+type defaultQuotaControl struct {
+	k8sGardenClient    kubernetes.Client
+	k8sGardenInformers gardeninformers.Interface
+}
+
+func (c *defaultQuotaControl) CheckQuota(shootObj *gardenv1beta1.Shoot, key string) error {
+	var (
+		clusterLifeTime     *int
+		quotaReferences     []v1.ObjectReference
+		shootExpirationTime time.Time
+
+		operationID = utils.GenerateRandomString(8)
+		shoot       = shootObj.DeepCopy()
+		shootLogger = logger.NewShootLogger(logger.Logger, shoot.Name, shoot.Namespace, operationID)
+	)
+
+	switch shoot.Spec.Cloud.SecretBindingRef.Kind {
+	case "PrivateSecretBinding":
+		psb, err := c.k8sGardenInformers.
+			PrivateSecretBindings().
+			Lister().
+			PrivateSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return err
+		}
+		quotaReferences = psb.Quotas
+	case "CrossSecretBinding":
+		csb, err := c.k8sGardenInformers.
+			CrossSecretBindings().
+			Lister().
+			CrossSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return err
+		}
+		quotaReferences = csb.Quotas
+	default:
+		return fmt.Errorf("Unknown Binding type %s", shoot.Spec.Cloud.SecretBindingRef.Kind)
+	}
+
+	for _, quotaRef := range quotaReferences {
+		quota, err := c.k8sGardenInformers.
+			Quotas().
+			Lister().
+			Quotas(quotaRef.Namespace).
+			Get(quotaRef.Name)
+		if err != nil {
+			return err
+		}
+
+		if quota.Spec.ClusterLifetimeDays == nil {
+			continue
+		}
+		if clusterLifeTime == nil || *quota.Spec.ClusterLifetimeDays < *clusterLifeTime {
+			clusterLifeTime = quota.Spec.ClusterLifetimeDays
+		}
+	}
+
+	if clusterLifeTime == nil {
+		return nil
+	}
+
+	if expirationTime, exists := shoot.Annotations[common.ClusterExpirationTime]; exists {
+		t, err := time.Parse(time.RFC3339, expirationTime)
+		if err != nil {
+			return err
+		}
+		shootExpirationTime = t.Add(time.Duration(*clusterLifeTime*24) * time.Hour)
+	} else {
+		annotations := shoot.Annotations
+		annotations[common.ClusterExpirationTime] = shoot.CreationTimestamp.Format(time.RFC3339)
+		shoot.Annotations = annotations
+
+		_, err := c.k8sGardenClient.
+			GardenClientset().
+			GardenV1beta1().
+			Shoots(shoot.Namespace).
+			Update(shoot)
+		if err != nil {
+			return err
+		}
+		shootExpirationTime = shoot.CreationTimestamp.Add(time.Duration(*clusterLifeTime*24) * time.Hour)
+	}
+
+	if time.Now().After(shootExpirationTime) {
+		shootLogger.Info("[SHOOT QUOTA] Shoot cluster lifetime expired. Shoot will be deleted.")
+
+		// We have to delete the shoot, because only apiserver can set a deletionTimestamp
+		err := c.k8sGardenClient.
+			GardenClientset().
+			GardenV1beta1().
+			Shoots(shoot.Namespace).
+			Delete(shoot.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		shootUpdated, err := c.k8sGardenClient.
```

Comment:
Can we write these statements into one line (line breakings are not used anywhere else in the code)?